File: android/ValleAPP/ValleTPVBack/src/main/java/com/valleapp/valletpv/Arqueo.java
package com.valleapp.valletpv;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.valleapp.valletpv.cashlogyActivitis.ArqueoCashlogyActivity;
import com.valleapp.valletpvlib.tools.HTTPRequest;
import com.valleapp.valletpvlib.tools.JSON;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;


public class Arqueo extends Activity {

    String server;
    LinearLayout pneGastos;
    LinearLayout pneEfectivo;
    TextView txtCambio;
    TextView txtGastos;
    TextView txtEfectivo;
    Double cambio =0.0, gastos=0.0, efectivo = 0.0;
    List<JSONObject> objGastos = new ArrayList<>();
    List<JSONObject> objEfectivo = new ArrayList<>();
    Context cx;


    @SuppressLint("HandlerLeak")
    private final Handler controller_http = new Handler(Looper.getMainLooper()) {
        @SuppressLint("DefaultLocale")
        public void handleMessage(Message msg) {
            String op = msg.getData().getString("op");
            String res = msg.getData().getString("RESPONSE");
            assert op != null;
            if (op.equals("cambio")) {
                try {
                    assert res != null;
                    JSONObject obj = new JSONObject(res);
                    cambio = obj.getDouble("cambio");
                    txtCambio.setText(String.format("%.2f €", cambio));
                    if (!obj.getBoolean("hay_arqueo")) {
                        mostrarMensaje("No hay Ticket para hacer un arqueo...\n " +
                                "Este arqueo remplaza al ultimo");
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }

            } else if(op.equals("arqueo")){
                assert res != null;
                if(res.equals("success")) finish();
                else{
                    mostrarMensaje("No hay Ticket para hacer un arqueo...\n " +
                            "Este arqueo remplaza al ultimo");
                }
            }
        }


    };

    private void mostrarMensaje(String men){
        Toast toast = new Toast(cx);
        @SuppressLint("InflateParams") View toast_view = LayoutInflater.from(cx).inflate(R.layout.texto_toast_simple, null);
        TextView textView = toast_view.findViewById(R.id.txt_label);
        textView.setTextSize(33);
        textView.setText(men);
        toast.setView(toast_view);
        toast.setDuration(Toast.LENGTH_LONG);
        toast.setGravity(Gravity.CENTER, 0, 80);
        toast.show();
    }

    @SuppressLint("DefaultLocale")
    private void rellenarGastos() {
        gastos=0.0;
        pneGastos.removeAllViews();

        for (JSONObject gasto : objGastos) {
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT);

            LayoutInflater inflater = (LayoutInflater) cx.getSystemService
                    (Context.LAYOUT_INFLATER_SERVICE);
            @SuppressLint("InflateParams") View v = inflater.inflate(R.layout.item_gastos, null);
            TextView can = v.findViewById(R.id.cantidad);
            TextView des = v.findViewById(R.id.Descripcion);
            ImageButton rm = v.findViewById(R.id.btn_borrar);
            rm.setTag(gasto);

            try {
            double cantidad = gasto.getDouble("Importe");
            String descrip = gasto.getString("Des");

            if (cantidad > 0 && descrip.length() > 0) {
                can.setText(String.format("%.2f €", cantidad));
                des.setText(descrip);
                gastos += cantidad;
                pneGastos.addView(v, params);
              }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        txtGastos.setText(String.format("%.2f €", gastos));

    }

    @SuppressLint("DefaultLocale")
    private void rellenarEfectivo(){

        efectivo = 0.0;
        pneEfectivo.removeAllViews();

        for(JSONObject e : objEfectivo) {
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.MATCH_PARENT);

            LayoutInflater inflater = (LayoutInflater) cx.getSystemService
                    (Context.LAYOUT_INFLATER_SERVICE);
            @SuppressLint("InflateParams") View v = inflater.inflate(R.layout.item_efectivo, null);
            TextView mon =  v.findViewById(R.id.txt_moneda);
            TextView can = v.findViewById(R.id.txt_cantidad);
            TextView tot = v.findViewById(R.id.Total);
            ImageButton rm = v.findViewById(R.id.btn_borrar);
            rm.setTag(e);

            try {
                double moneda = e.getDouble("Moneda");
                int cantidad = e.getInt("Can");
                mon.setText(String.format("%01.2f €", moneda));
                can.setText(String.format("%s", cantidad));
                tot.setText(String.format("%01.2f €", (cantidad * moneda)));
                efectivo += cantidad * moneda;
                pneEfectivo.addView(v, params);
            } catch (JSONException x) {
                x.printStackTrace();
            }
        }
        txtEfectivo.setText(String.format("%.2f €", efectivo));
    }

    public void clickAbrirCaja(View v){
        new HTTPRequest(server+"/impresion/abrircajon",new ContentValues(),"open", controller_http);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_arqueo);
        CargarPreferencias();
        pneGastos = findViewById(R.id.pneGastos);
        pneEfectivo = findViewById(R.id.pneEfectivo);
        txtCambio = findViewById(R.id.lblCambio);
        txtEfectivo = findViewById(R.id.lblEfectivo);
        txtGastos = findViewById(R.id.lblGastos);
        this.cx = this;
    }

    public void addEfectivo(View v){
       final Dialog dlg = new Dialog(this);
       dlg.setContentView(R.layout.add_efectivo);
       dlg.setTitle("Agregar efectivo");
       ImageButton s = dlg.findViewById(R.id.btn_salir_monedas);
       ImageButton ok = dlg.findViewById(R.id.btn_guardar_preferencias);
       final TextView m = dlg.findViewById(R.id.txtMoneda);
       final TextView c = dlg.findViewById(R.id.txtCantidad);
          s.setOnClickListener(view -> dlg.cancel());
          ok.setOnClickListener(view -> {
            try {
                Double moneda = Double.parseDouble(m.getText().toString().replace(",", "."));
                int cantidad = Integer.parseInt(c.getText().toString());
                if ((moneda * cantidad) > 0) {
                    try {
                        JSONObject obj = new JSONObject();
                        obj.put("Can", cantidad);
                        obj.put("Moneda", moneda);
                        objEfectivo.add(obj);
                        rellenarEfectivo();
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
                dlg.cancel();
            }catch(Exception exp){
              exp.printStackTrace();
            }
          });
          dlg.show();
      }

      public void addGastos(View v){
          final Dialog dlg = new Dialog(this);
          dlg.setContentView(R.layout.add_gastos);
          dlg.setTitle("Agregar gasto");
          ImageButton s = dlg.findViewById(R.id.Salir);
          ImageButton ok = dlg.findViewById(R.id.Aceptar);
          final TextView txtDes = dlg.findViewById(R.id.txtDescripcion);
          final TextView imp = dlg.findViewById(R.id.txtImporte);
          s.setOnClickListener(view -> dlg.cancel());
          ok.setOnClickListener(view -> {
              try {
                  Double Importe = Double.parseDouble(imp.getText().toString().replace(",", "."));
                  String des = txtDes.getText().toString();
                  if(Importe>0 &&  des.length()>0){
                          JSONObject obj = new JSONObject();
                          obj.put("Des",des);
                          obj.put("Importe", Importe);
                          objGastos.add(obj);
                          rellenarGastos();

                  }
                  dlg.cancel();
              } catch (Exception e) {
                  e.printStackTrace();
              }
          });
          dlg.show();
      }

      public void EditCambio(View v){
          final Dialog dlg = new Dialog(this);
          dlg.setContentView(R.layout.edit_cambio);
          dlg.setTitle("Editar Cambio");
          ImageButton s = dlg.findViewById(R.id.salirCambio);
          ImageButton ok =  dlg.findViewById(R.id.aceptarCam);
          final TextView txtDes = dlg.findViewById(R.id.cambio);
          s.setOnClickListener(view -> dlg.cancel());
          ok.setOnClickListener(view -> {
            try{
              cambio = Double.parseDouble(txtDes.getText().toString());
              txtCambio.setText(String.format("%s €", cambio));
              dlg.cancel();
            } catch (Exception e) {
                e.printStackTrace();
            }
          });
         dlg.show();
      }

      public void arquearCaja(View v){
           if (cambio>0) {
               ContentValues p = new ContentValues();
               p.put("cambio", Double.toString(cambio));
               p.put("efectivo", Double.toString(efectivo));
               p.put("gastos", Double.toString(gastos));
               p.put("des_efectivo", objEfectivo.toString());
               p.put("des_gastos", objGastos.toString());
               new HTTPRequest(server + "/arqueos/arquear", p, "arqueo", controller_http);
           }else{
               mostrarMensaje("El cambio debe ser mayor que 0 €");
           }

      }

    public void CargarPreferencias() {
        JSON json = new JSON();
        try {
            JSONObject pref = json.deserializar("preferencias.dat", this);
            if (pref == null) {
                Intent intent = new Intent(this, PreferenciasTPV.class);
                startActivity(intent);
            } else {
                server = pref.getString("URL");

                // Verifica si la preferencia indica que se debe usar Cashlogy
                boolean usaCashlogy = pref.getBoolean("usaCashlogy");

                if (usaCashlogy) {
                    // Solicita los datos del servidor primero
                    new HTTPRequest(server + "/arqueos/getcambio", new ContentValues(), "cambio", new Handler(Looper.getMainLooper()) {
                        @Override
                        public void handleMessage(Message msg) {
                            String res = msg.getData().getString("RESPONSE");
                            if (res != null) {
                                try {
                                    JSONObject obj = new JSONObject(res);
                                    // Recoge los datos del servidor
                                    double cambio = obj.getDouble("cambio");
                                    boolean hayArqueo = obj.getBoolean("hay_arqueo");
                                    double stacke = obj.getDouble("stacke");
                                    double cambio_real = obj.getDouble("cambio_real");

                                    // Inicia ArqueoCashlogyActivity pasando los datos
                                    Intent intent = new Intent(Arqueo.this, ArqueoCashlogyActivity.class);
                                    intent.putExtra("cambio", cambio);
                                    intent.putExtra("hayArqueo", hayArqueo);
                                    intent.putExtra("stacke", stacke);
                                    intent.putExtra("cambio_real", cambio_real);

                                    // Añadir preferencias al Intent
                                    intent.putExtra("URL", server);
                                    intent.putExtra("URL_Cashlogy", pref.getString("URL_Cashlogy"));
                                    intent.putExtra("usaCashlogy", pref.getBoolean("usaCashlogy"));

                                    startActivity(intent);
                                    finish(); // Finaliza la actividad actual
                                } catch (JSONException e) {
                                    e.printStackTrace();
                                }
                            }
                        }
                    });
                } else {
                    // Si no se usa Cashlogy, continúa con la lógica normal
                    new HTTPRequest(server + "/arqueos/getcambio", new ContentValues(), "cambio", controller_http);
                }
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }




    public void clickBorrarEfc(View v){
        JSONObject obj = (JSONObject)v.getTag();
        objEfectivo.remove(obj);
        rellenarEfectivo();
    }

    public void clickBorrarGasto(View v){
        JSONObject obj = (JSONObject)v.getTag();
        objGastos.remove(obj);
        rellenarGastos();
    }

}

File: android/ValleTPV/ValleTPV/src/main/java/com/valleapp/valletpv/Arqueo.kt
package com.valleapp.valletpv

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.content.ContentValues
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.os.Message
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import com.tu.paquete.CustomToast
import com.valleapp.valletpv.cashlogyActivitis.ArqueoCashlogyActivity
import com.valleapp.valletpv.tools.ServiceTPV
import com.valleapp.valletpvlib.comunicacion.HTTPRequest
import com.valleapp.valletpvlib.tools.JSON
import org.json.JSONException
import org.json.JSONObject

class Arqueo : Activity() {

    private var myServicio: ServiceTPV? = null

    private val mConexion = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, iBinder: IBinder?) {
            myServicio = (iBinder as ServiceTPV.MyBinder).service
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            myServicio = null
        }
    }

    private lateinit var server: String
    private lateinit var pneGastos: LinearLayout
    private lateinit var pneEfectivo: LinearLayout
    private lateinit var txtCambio: TextView
    private lateinit var txtGastos: TextView
    private lateinit var txtEfectivo: TextView
    private var cambio = 0.0
    private var gastos = 0.0
    private var efectivo = 0.0
    private val objGastos = ArrayList<JSONObject>()
    private val objEfectivo = ArrayList<JSONObject>()
    private lateinit var cx: Context
    private val custuomToast = CustomToast(this)

    @SuppressLint("HandlerLeak")
    private val controllerHttp = object : Handler(Looper.getMainLooper()) {
        @SuppressLint("DefaultLocale")
        override fun handleMessage(msg: Message) {
            val op = msg.data.getString("op")
            val res = msg.data.getString("RESPONSE")
            if (op == null || res == null) return
            if (op == "cambio") {
                try {
                    val obj = JSONObject(res)
                    cambio = obj.getDouble("cambio")
                    txtCambio.text = String.format("%.2f €", cambio)
                    if (!obj.getBoolean("hay_arqueo")) {
                        custuomToast.showCenter(
                            "No hay Ticket para hacer un arqueo...\n " +
                                    "Este arqueo remplaza al ultimo"
                        )
                    }
                } catch (e: JSONException) {
                    Log.e("ARQUEO_ERR", e.toString())
                }
            } else if (op == "arqueo") {
                try {
                    val obj = JSONObject(res)
                    if (obj.getBoolean("success")) finish()
                    else {
                        mostrarMensaje(
                            "No hay Ticket para hacer un arqueo...\n " +
                                    "Este arqueo remplaza al ultimo"
                        )
                    }
                } catch (e: Exception) {
                    Log.e("ARQUEO_ERR", e.toString())
                }
            }
        }
    }

    @SuppressLint("DefaultLocale")
    private fun rellenarGastos() {
        gastos = 0.0
        pneGastos.removeAllViews()

        for (gasto in objGastos) {
            val params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )

            val inflater = cx.getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            @SuppressLint("InflateParams")
            val v = inflater.inflate(R.layout.item_gastos, null)
            val can = v.findViewById<TextView>(R.id.cantidad)
            val des = v.findViewById<TextView>(R.id.Descripcion)
            val rm = v.findViewById<ImageButton>(R.id.btn_borrar)
            rm.tag = gasto

            try {
                val cantidad = gasto.getDouble("Importe")
                val descrip = gasto.getString("Des")

                if (cantidad > 0 && descrip.isNotEmpty()) {
                    can.text = String.format("%.2f €", cantidad)
                    des.text = descrip
                    gastos += cantidad
                    pneGastos.addView(v, params)
                }
            } catch (e: JSONException) {
                Log.e("ARQUEO_ERR", e.toString())
            }
        }
        txtGastos.text = String.format("%.2f €", gastos)
    }

    @SuppressLint("DefaultLocale")
    private fun rellenarEfectivo() {
        efectivo = 0.0
        pneEfectivo.removeAllViews()

        for (e in objEfectivo) {
            val params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
            )

            val inflater = cx.getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            @SuppressLint("InflateParams")
            val v = inflater.inflate(R.layout.item_efectivo, null)
            val mon = v.findViewById<TextView>(R.id.txt_moneda)
            val can = v.findViewById<TextView>(R.id.txt_cantidad)
            val tot = v.findViewById<TextView>(R.id.Total)
            val rm = v.findViewById<ImageButton>(R.id.btn_borrar)
            rm.tag = e

            try {
                val moneda = e.getDouble("Moneda")
                val cantidad = e.getInt("Can")
                mon.text = String.format("%01.2f €", moneda)
                can.text = String.format("%s", cantidad)
                tot.text = String.format("%01.2f €", (cantidad * moneda))
                efectivo += cantidad * moneda
                pneEfectivo.addView(v, params)
            } catch (x: JSONException) {
                Log.e("ARQUEO_ERR", x.toString())
            }
        }
        txtEfectivo.text = String.format("%.2f €", efectivo)
    }

    fun clickAbrirCaja(v: View) {
        val p = ContentValues()
        p.put("uid", myServicio?.getUID())
        HTTPRequest("$server/impresion/abrircajon", p, "open", controllerHttp)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_arqueo)

        // Enlazar con el servicio
        val intent = Intent(this, ServiceTPV::class.java)
        bindService(intent, mConexion, BIND_AUTO_CREATE)

        cargarPreferencias()
        pneGastos = findViewById(R.id.pneGastos)
        pneEfectivo = findViewById(R.id.pneEfectivo)
        txtCambio = findViewById(R.id.lblCambio)
        txtEfectivo = findViewById(R.id.lblEfectivo)
        txtGastos = findViewById(R.id.lblGastos)
        this.cx = this
        val btnSalir = findViewById<ImageButton>(R.id.btnSalir)
        btnSalir.setOnClickListener { finish() }
    }

    fun addEfectivo(v: View) {
        val dlg = Dialog(this)
        dlg.setContentView(R.layout.add_efectivo)
        dlg.setTitle("Agregar efectivo")
        val s = dlg.findViewById<ImageButton>(R.id.btn_salir_monedas)
        val ok = dlg.findViewById<ImageButton>(R.id.btn_guardar_preferencias)
        val m = dlg.findViewById<TextView>(R.id.txtMoneda)
        val c = dlg.findViewById<TextView>(R.id.txtCantidad)
        s.setOnClickListener { dlg.cancel() }
        ok.setOnClickListener {
            try {
                val moneda = m.text.toString().replace(",", ".").toDouble()
                val cantidad = c.text.toString().toInt()
                if ((moneda * cantidad) > 0) {
                    try {
                        val obj = JSONObject()
                        obj.put("Can", cantidad)
                        obj.put("Moneda", moneda)
                        objEfectivo.add(obj)
                        rellenarEfectivo()
                    } catch (e: JSONException) {
                        Log.e("ARQUEO_ERR", e.toString())
                    }
                }
                dlg.cancel()
            } catch (exp: Exception) {
                Log.e("ARQUEO_ERR", exp.toString())
            }
        }
        dlg.show()
    }

    fun addGastos(v: View) {
        val dlg = Dialog(this)
        dlg.setContentView(R.layout.add_gastos)
        dlg.setTitle("Agregar gasto")
        val s = dlg.findViewById<ImageButton>(R.id.Salir)
        val ok = dlg.findViewById<ImageButton>(R.id.Aceptar)
        val txtDes = dlg.findViewById<TextView>(R.id.txtDescripcion)
        val imp = dlg.findViewById<TextView>(R.id.txtImporte)
        s.setOnClickListener { dlg.cancel() }
        ok.setOnClickListener {
            try {
                val importe = imp.text.toString().replace(",", ".").toDouble()
                val des = txtDes.text.toString()
                if (importe > 0 && des.isNotEmpty()) {
                    val obj = JSONObject()
                    obj.put("Des", des)
                    obj.put("Importe", importe)
                    objGastos.add(obj)
                    rellenarGastos()
                }
                dlg.cancel()
            } catch (e: Exception) {
                Log.e("ARQUEO_ERR", e.toString())
            }
        }
        dlg.show()
    }

    fun editCambio(v: View) {
        val dlg = Dialog(this)
        dlg.setContentView(R.layout.edit_cambio)
        dlg.setTitle("Editar Cambio")
        val s = dlg.findViewById<ImageButton>(R.id.salirCambio)
        val ok = dlg.findViewById<ImageButton>(R.id.aceptarCam)
        val txtDes = dlg.findViewById<TextView>(R.id.cambio)
        s.setOnClickListener { dlg.cancel() }
        ok.setOnClickListener {
            try {
                cambio = txtDes.text.toString().toDouble()
                txtCambio.text = String.format("%s €", cambio)
                dlg.cancel()
            } catch (e: Exception) {
                Log.e("ARQUEO_ERR", e.toString())
            }
        }
        dlg.show()
    }

    fun arquearCaja(v: View) {
        if (cambio > 0) {
            val p = ContentValues()
            p.put("cambio", cambio.toString())
            p.put("efectivo", efectivo.toString())
            p.put("gastos", gastos.toString())
            p.put("des_efectivo", objEfectivo.toString())
            p.put("des_gastos", objGastos.toString())
            p.put("uid", myServicio?.getUID())
            HTTPRequest("$server/arqueos/arquear", p, "arqueo", controllerHttp)
        } else {
            mostrarMensaje("El cambio debe ser mayor que 0 €")
        }
    }

    fun cargarPreferencias() {
        val json = JSON()
        try {
            val pref = json.deserializar("preferencias.dat", this)
            if (pref == null) {
                val intent = Intent(this, PreferenciasTPV::class.java)
                startActivity(intent)
            } else {
                server = pref.getString("URL")

                // Verifica si la preferencia indica que se debe usar Cashlogy
                val usaCashlogy = pref.getBoolean("usaCashlogy")

                if (usaCashlogy) {
                    // Solicita los datos del servidor primero
                    val p = ContentValues()
                    p.put("uid", myServicio?.getUID())
                    HTTPRequest("$server/arqueos/getcambio", p, "cambio", object : Handler(Looper.getMainLooper()) {
                        override fun handleMessage(msg: Message) {
                            val res = msg.data.getString("RESPONSE")
                            if (res != null) {
                                try {
                                    val obj = JSONObject(res)
                                    // Recoge los datos del servidor
                                    val cambio = obj.getDouble("cambio")
                                    val hayArqueo = obj.getBoolean("hay_arqueo")
                                    val stacke = obj.getDouble("stacke")
                                    val cambioReal = obj.getDouble("cambio_real")

                                    // Inicia ArqueoCashlogyActivity pasando los datos
                                    val intent = Intent(this@Arqueo, ArqueoCashlogyActivity::class.java)
                                    intent.putExtra("cambio", cambio)
                                    intent.putExtra("hayArqueo", hayArqueo)
                                    intent.putExtra("stacke", stacke)
                                    intent.putExtra("cambio_real", cambioReal)

                                    // Añadir preferencias al Intent
                                    intent.putExtra("URL", server)
                                    intent.putExtra("URL_Cashlogy", pref.getString("URL_Cashlogy"))
                                    intent.putExtra("usaCashlogy", pref.getBoolean("usaCashlogy"))

                                    startActivity(intent)
                                    finish() // Finaliza la actividad actual
                                } catch (e: JSONException) {
                                    Log.e("ARQUEO_ERR", e.toString())
                                }
                            }
                        }
                    })
                } else {
                    // Si no se usa Cashlogy, continúa con la lógica normal
                    val p = ContentValues()
                    p.put("uid", myServicio?.getUID())
                    HTTPRequest("$server/arqueos/getcambio", p, "cambio", controllerHttp)
                }
            }
        } catch (e: Exception) {
            Log.e("ARQUEO_ERR", e.toString())
        }
    }

    fun clickBorrarEfc(v: View) {
        val obj = v.tag as JSONObject
        objEfectivo.remove(obj)
        rellenarEfectivo()
    }

    fun clickBorrarGasto(v: View) {
        val obj = v.tag as JSONObject
        objGastos.remove(obj)
        rellenarGastos()
    }

    private fun mostrarMensaje(texto: String) {
        // Este método no está implementado en el código Java original,
        // asumo que debería mostrar un Toast o similar
        custuomToast.showCenter(texto)
    }

    override fun onDestroy() {
        super.onDestroy()
        // Desenlazar del servicio
        myServicio?.let { unbindService(mConexion) }
    }
}